### PR TITLE
ESQL: Speed up grouping by bytes

### PR DESCRIPTION
**Pull request by nik9000**
This speeds up grouping by bytes valued fields (keyword, text, ip, and wildcard) when the input is an ordinal vector:
```
    bytes_refs 22.213 ± 0.322 -> 19.848 ± 0.205 ns/op (*maybe* real, maybe noise. still good)
       ordinal didn't exist   ->  2.988 ± 0.011 ns/op
```
I see this as 20ns -> 3ns, an 85% speed up. We never hard the ordinals branch before so I'm expecting the same performance there - about 20ns per op.

This also speeds up grouping by a pair of byte valued fields:
```
two_bytes_refs 83.112 ± 42.348  -> 46.521 ± 0.386 ns/op
  two_ordinals 83.531 ± 23.473  ->  8.617 ± 0.105 ns/op
```
The speed up is much better when the fields are ordinals because hashing bytes is comparatively slow.

I expect that you'll also see a speed up in the case of three ordinals though I didn't add a microbenchmark for it.

I believe the ordinals case is quite common. I've run into it in quite a few profiles.

Note: The actual ordinals speed up is based on the size of the dictionary. This test has a five element dictionary. That's *normal* for an `enum` style column. But not for a `transaction_id` style column. You'll see a worse speed up there - but it'll still be substantial.

### CODE REVIEW

File: x-pack/plugin/esql/compute/src/main/generated-src/org/elasticsearch/compute/aggregation/blockhash/BytesRefBlockHash.java [hunk 1]
```diff
@@ -132,9 +153,11 @@ private IntBlock addOrdinalsBlock(OrdinalBytesRefBlock inputBlock) {
                         builder.appendInt(ord);
                     }
                     default -> {
+                        int start = firstIndex;
+                        int end = firstIndex + valueCount;
                         builder.beginPositionEntry();
-                        for (int v = 0; v < valueCount; v++) {
-                            int ord = hashOrds.getInt(inputOrds.getInt(firstIndex + i));
+                        for (int i = start; i < end; i++) {
+                            int ord = hashOrds.getInt(inputOrds.getInt(i));
```
[the remark this answers is not in the document]
Comment (nik9000) [reply x-pack/plugin/esql/compute/src/main/generated-src/org/elasticsearch/compute/aggregation/blockhash/BytesRefBlockHash.java:161]: I don't think we can land here, but we come *super* close. The only way to get ordinals at the moment is with `BlockDocValuesReader.SingletonOrdinals` and that only loads from lucene when the field is single valued. This path only works for `Blocks`. Specifically, this arm is for multi-valued positions. Which we aren't loading.

File: x-pack/plugin/esql/compute/src/test/java/org/elasticsearch/compute/aggregation/blockhash/BlockHashTests.java [hunk 2]
```diff
@@ -460,6 +463,133 @@ public void testBytesRefHashWithMultiValuedFields() {
...
+                    assertThat(ordsAndKeys.description, startsWith("PackedValuesBlockHash{groups=[0:BYTES_REF], entries=4, size="));
+                    assertThat(ordsAndKeys.description, endsWith("b}"));
+                    assertOrds(
+                        ordsAndKeys.ords,
+                        new int[] { 0 },
+                        new int[] { 0, 1 },
+                        new int[] { 1, 2 },
+                        new int[] { 2, 1 },
+                        new int[] { 3 },
+                        new int[] { 2, 1 }
```
[the remark this answers is not in the document]
Comment (nik9000) [reply x-pack/plugin/esql/compute/src/test/java/org/elasticsearch/compute/aggregation/blockhash/BlockHashTests.java:571]: `BlockHash` is designed to be pushed into grouping aggs. That's really what it's for. And aggs have "does this row land in this bucket" semantics, not "how many times does this row land in this bucket". It's a set of buckets the row lands in, not a list.

File: x-pack/plugin/esql/compute/src/main/generated-src/org/elasticsearch/compute/aggregation/blockhash/BytesRefBlockHash.java [hunk 3]
```diff
@@ -113,15 +121,29 @@ public ReleasableIterator<IntBlock> lookup(Page page, ByteSizeValue targetBlockS
...
+                int ord = hashOrds.getInt(inputOrds.getInt(p));
+                builder.appendInt(ord);
+            }
+            return builder.build();
+        }
+    }
+
+    private IntBlock addOrdinalsBlock(OrdinalBytesRefBlock inputBlock) {
+        try (
+            IntBlock inputOrds = new MultivalueDedupeInt(inputBlock.getOrdinalsBlock()).dedupeToBlockAdaptive(blockFactory);
```
[the remark this answers is not in the document]
Comment (nik9000) [reply x-pack/plugin/esql/compute/src/main/generated-src/org/elasticsearch/compute/aggregation/blockhash/BytesRefBlockHash.java:140]: Yeah. The contacts for BlockHashes is to deduplicate.

File: x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/aggregation/blockhash/BytesRef2BlockHash.java [hunk 4]
```diff
@@ -0,0 +1,196 @@
...
+import org.elasticsearch.compute.data.IntBlock;
+import org.elasticsearch.compute.data.IntVector;
+import org.elasticsearch.compute.data.Page;
+import org.elasticsearch.core.ReleasableIterator;
+import org.elasticsearch.core.Releasables;
+
+import java.util.Locale;
+
+/**
+ * Maps three {@link BytesRefBlock}s to group ids.
```
[the remark this answers is not in the document]
Comment (nik9000) [reply x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/aggregation/blockhash/BytesRef2BlockHash.java]: :+1:

**Comment by nik9000** [comment]
Here's an apples to apples comparison on some real data:
```
Before:
          {
            "operator" : "HashAggregationOperator[blockHash=PackedValuesBlockHash{groups=[0:BYTES_REF, 1:BYTES_REF], entries=8663014, size=767.9mb}, aggregators=[GroupingAggregator[aggregatorFunction=CountGroupingAggregatorFunction[channels=[]], mode=INITIAL]]]",
            "status" : {
              "hash_nanos" : 5674441635,
              "aggregation_nanos" : 49922360,
              "pages_processed" : 3985
            }
          },
          {
            "operator" : "HashAggregationOperator[blockHash=PackedValuesBlockHash{groups=[0:BYTES_REF, 1:BYTES_REF], entries=8663014, size=767.9mb}, aggregators=[GroupingAggregator[aggregatorFunction=CountGroupingAggregatorFunction[channels=[2, 3]], mode=FINAL]]]",
            "status" : {
              "hash_nanos" : 4839721641,
              "aggregation_nanos" : 64996752,
              "pages_processed" : 1
            }
          },

After:
          {
            "operator" : "HashAggregationOperator[blockHash=BytesRef2BlockHash{keys=[channel1=0, channel2=1], entries=8662908}, aggregators=[GroupingAggregator[aggregatorFunction=CountGroupingAggregatorFunction[channels=[]], mode=INITIAL]]]",
            "status" : {
              "hash_nanos" : 2807552505,
              "aggregation_nanos" : 45913603,
              "pages_processed" : 3985
            }
          },
          {
            "operator" : "HashAggregationOperator[blockHash=BytesRef2BlockHash{keys=[channel1=0, channel2=1], entries=8662908}, aggregators=[GroupingAggregator[aggregatorFunction=CountGroupingAggregatorFunction[channels=[2, 3]], mode=FINAL]]]",
            "status" : {
              "hash_nanos" : 2368230273,
              "aggregation_nanos" : 73475991,
              "pages_processed" : 1
            }
          },
```

That's about 50% faster hashing. And in this case hashing is dominating the query. So I'm quite hopeful this'll translate well.

File: x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/aggregation/blockhash/BooleanBlockHash.java [hunk 5]
```diff
@@ -25,8 +25,9 @@
 import static org.elasticsearch.compute.operator.mvdedupe.MultivalueDedupeBoolean.TRUE_ORD;
 
 /**
- * Maps a {@link BooleanBlock} column to group ids. Assigns group
- * {@code 0} to {@code false} and group {@code 1} to {@code true}.
+ * Maps a {@link BooleanBlock} column to group ids. Assigns
+ * {@code 0} to {@code null}, {@code 1} to {@code false}, and
+ * {@code 1} to {@code true}.
```
Comment (ivancea) [inline x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/aggregation/blockhash/BooleanBlockHash.java:30]:
```suggestion
 * {@code 2} to {@code true}.
```

**Review by ivancea — APPROVED** [review]
LGTM

**Comment by nik9000** [comment]
@elasticmachine, test this please

**Pull request merged**